### DISCUSSION TITLE
option to fill nodata in raster

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#84] [General]
When processing e.g. CHMs /DTMs we might face areas/ pixels with nodata. 

I am wondering if it would make sense to add a focal filtering stage which would allow e.g. to apply a median focal filter just on nodata areas. 
There is `pit_fill` which does something similar, however there is no option run it only on nodata and hence it smoothes the entire raster.
We could also use a TIN but this also kind of smoothes the ratser and is probably not wanted. 
``` r
library(lasR)
f <- system.file("extdata", "Topography.las", package="lasR")

read = reader_las(filter = "-keep_first -keep_class 2 -filter_and")
del1 = triangulate(filter = keep_ground())
norm = transform_with(del1, "-")
chm1 = rasterize(2, "max")
pit = pit_fill(chm1)
del2 = triangulate(filter = keep_first())
chm2 = rasterize(2, del2)
pipeline <- read + del1 + norm + chm1 + pit + del2 + chm2 
ans <- exec(pipeline, on = f, with = list(ncores = concurrent_points(half_cores())))

terra::plot(ans[[1]])
```

![](https://i.imgur.com/AhxuFd2.png)

``` r
terra::plot(ans[[2]])
```

![](https://i.imgur.com/sS455Y0.png)

``` r
terra::plot(ans[[3]])
```

![](https://i.imgur.com/wzKr5Nt.png)

**Comment by Jean-Romain** [comment]
I think it is important to have no data. It shows users what they are currently doing such as requesting a too fine grain output with a too low density point cloud. Also `lasR` like `lidR` are point cloud processing software I'm not keen too reinvent the wheel too much with raster features. 

Anyway I planned to add a focal function for raster to smooth the canopy and improve tree detection and segmentation. I could add a function to fill NAs.

**Reply by wiesehahn** [reply]
> I think it is important to have no data. It shows users what they are currently doing such as requesting a too fine grain output with a too low density point cloud. 

Maybe, but a focal filter with small kernel would not fill the water areas above but simply fill single NA pixels which are then assumed to be as high as average sourrounding. 

> Also `lasR` like `lidR` are point cloud processing software I'm not keen too reinvent the wheel too much with raster features.

Alright, I understand that, could be done with gdal as well.

**Reply by Jean-Romain** [reply]
> Alright, I understand that, could be done with gdal as well.

Ideally yes post prod should be done with `terra` or another `GIS` software. But sometime post-prod is required to feed next stages and `lasR` will try to cover these cases when possible. I also plan to add a raster processing engine to work with raster without the point cloud

```r
chm = reader_raster()
lmf =  local_maximum_raster(chm)
seg  = region_growning(chm, lmf)
pipeline = chm + lmf + seg
```

**Comment by Jean-Romain** [comment]
## v0.10.0

``` r
library(lasR)
f <- system.file("extdata", "Topography.las", package = "lasR")

chm = rasterize(2, "zmax")
chm2 = lasR:::focal(chm, 8, fun = "mean")
chm3 = lasR:::focal(chm, 8, fun = "max")
pipeline <- reader_las() + chm + chm2 + chm3
u = exec(pipeline, on = f)

terra::plot(u[[1]])
```

![](https://i.imgur.com/N8iRu9k.png)

``` r
terra::plot(u[[2]])
```

![](https://i.imgur.com/E3Yn6ji.png)

``` r
terra::plot(u[[3]])
```

![](https://i.imgur.com/8UCzTtU.png)

<sup>Created on 2024-08-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by wiesehahn** [comment]
For my future self:
Its simple and efficient as following:
``` r
library(lasR)
#> Warning: package 'lasR' was built under R version 4.4.2
f <- system.file("extdata", "Topography.las", package = "lasR")

chm = rasterize(2, "zmax")
pipeline <- reader_las() + chm
ans = exec(pipeline, on = f)
terra::plot(ans)
```

![](https://i.imgur.com/tGvz1N3.png)

``` r
gdalraster::fillNodata(terra::sources(ans), max_dist = 3, band=1)
#> 0...10...20...30...40...50...60...70...80...90...100 - done.
terra::plot(ans)
```

![](https://i.imgur.com/uRfDjIk.png)

<sup>Created on 2024-12-07 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>
